### ISSUE TITLE
[Submission] 36 params - Tom Bukic

### ISSUE THREAD

**Issue by tbukic** [anadim/AdderBoard#82]
### Author

tbukic

### Unique Parameter Count

36

### Accuracy

100.00% on 10,010 tests (verify.py seed=2025: 10K random + 10 edge cases)

### Method

Trained (SGD/Adam)

### Architecture

1L decoder-only transformer + circular arc embedding, d=3, 1h/1kv, hd=4, ff=2, RoPE theta=3, SwiGLU, RMSNorm

### Key Tricks

- Circular arc embedding (3 params instead of 30): digit tokens mapped via A*cos(start + stride*i)
- K = rotation(Q) (1-param 2D rotation in head_dim space replaces 12-param key projection)
- V = Q (value shares query weights exactly, 0 extra params — replaces 12-param value projection)
- Tied O=Q^T (output projection = transpose of query projection, saves 12 params)
- All 3 RMSNorms shared (ln1, ln2, final_norm share weights, saves 6 params)
- Tied QK norms (q_norm and k_norm share weights, saves 4 params)
- down = rotation(up^T) (1-param rotation + transpose replaces 6-param down projection)
- Tied lm_head to dynamic embedding table
- RoPE positional encoding (zero trainable params, theta=3)
- SwiGLU MLP at ff=2
- Grokfast-EMA gradient filter (alpha=0.98, lambda=3.0)
- Iterated targeted fine-tuning (2 rounds, 70 cumulative error pairs, lr=1e-4)

### Link to Code

https://github.com/tbukic/M10S-Transformer/blob/main/submissions/submission_36p.py

### Verification Output

```shell
Model: M10S-36p
Author: Tom Bukic
Parameters (unique): 36
Architecture: 1L decoder-only transformer + circular arc embed, d=3, 1h/1kv, hd=4, ff=2, K=rot(Q), V=Q, tieQO, shnorm, tieQKnorm, down=rot(up^T)

Results: 10010/10010 correct (100.00%)
Status: QUALIFIED (threshold: 99%)
```

### Additional Notes

- Parameter budget (36 total):
  - 3 circular arc embedding (A, start, stride)
  - 3 ln1 RMSNorm (shared with ln2 and final_norm via shnorm)
  - 1 k_rotation angle (K = rotation(Q), 2D rotation in head_dim=4 space)
  - 12 q_proj (shared with v_proj as V=Q, and o_proj via tieQO)
  - 4 q_norm (shared with k_norm via tieQKnorm)
  - 6 up_proj (shared with down_proj via rotation+transpose)
  - 1 down_proj rotation angle (down = rotation(up^T))
  - 6 gate_proj
  Total: 3 + 3 + 1 + 12 + 4 + 6 + 1 + 6 = 36
- 100% accuracy — 0 errors on 10,010 verify.py samples, 0 errors on 50K independent holdout (seed=99)
- - Training pipeline: seed=8, 300K cosine LR sweep (curr+gf+awd) → 365K Grokfast-EMA push (lr=5e-5, lambda=3.0) to 99.52% → 2 rounds iterated targeted FT (70 pairs, lr=1e-4, no WD) → 0 errors
- The down=rotation(up^T) trick saves 5 params vs the 41p architecture (6-param down_proj → 1-param rotation angle)
- Created with SuperchargeAI + Claude Code

**Parameter Oracle Audit (param_count.py from [#41](https://github.com/anadim/AdderBoard/issues/41)):**
```
submission: submissions/submission_36p.py
metadata params: 36

legacy counts
params (alias tying):          36
params (raw):                  42
params (value dedup, legacy):  36
float buffers (alias tying):   1440
float buffers (value dedup):   1440
effective (alias+buffers):     1476
effective (value+buffers):     1476

strict oracle counts
all buffers (alias tying):     1440
all buffers (raw):             1440
all buffers (value dedup):     1440
unregistered tensor-like (alias):  0
unregistered tensor-like (raw):    0
unregistered tensor-like (value):  0
python numeric state:          48
source numeric literals:       8
source unique numerics:        5

module/global extras
module tensor-like (alias):    0
module tensor-like (raw):      0
module tensor-like (value):    0
module python numeric state:   2

strict tensor oracle (alias):  1476
strict tensor oracle (raw):    1482
strict state oracle:           1530
strict source oracle:          1538

strict global tensor (raw):    1482
strict global state oracle:    1532
strict global source oracle:   1540

deltas
metadata -> alias delta:       0
alias -> strict tensor delta:  1446
strict tensor -> source delta: 56
alias -> strict global delta:  1496
```

Note on buffers (1440 float): these are **precomputed RoPE cosine/sine tables** (72+72=144 floats) and a **causal attention mask** (36×36=1296 floats) stored via `register_buffer()`. They are deterministic constants computed from fixed hyperparameters (theta=3.0, sequence length) — equivalent to sinusoidal positional encodings, which the AdderBoard rules explicitly exclude from parameter counts. The same values could be computed on-the-fly each forward pass (as some other submissions do), eliminating all buffers from the oracle report. We chose `register_buffer` for minor inference speed, not to hide parameters.

The 5 unique source numerics are Python boilerplate: `0` (path insert / indexing), `1` (tensor indexing), `10` (decimal conversion in `10**i`), `36` (metadata param count), and `True` (`weights_only` flag). None are trained values or encode hidden information — the oracle scans only the submission file itself, not transitively imported modules.
